protomap/proto.go

```diff
@@ -544,90 +548,148 @@ func ProtoFromPaths(p proto.Message, vals map[*gpb.Path]interface{}, opt ...Unma
 		return np
 	}
 
-	// directCh is a map between the absolute schema path for a particular value, and
-	// the value specified.
-	directCh := map[*gpb.Path]interface{}{}
-	for p, v := range vals {
-		absPath := &gpb.Path{
-			Elem: append(append([]*gpb.PathElem{}, schemaPath(valPrefix).Elem...), p.Elem...),
-		}
+	findChildren := func(vals map[*gpb.Path]any, valPrefix *gpb.Path, protoPrefix *gpb.Path, directOnly, mustBeChildren bool) (map[*gpb.Path]any, error) {
+		// directCh is a map between the absolute schema path for a particular value, and
+		// the value specified.
+		directCh := map[*gpb.Path]interface{}{}
+		for p, v := range vals {
+			absPath := &gpb.Path{
+				Elem: append(append([]*gpb.PathElem{}, schemaPath(valPrefix).Elem...), p.Elem...),
+			}
 
-		if !util.PathMatchesPathElemPrefix(absPath, protoPrefix) {
-			return fmt.Errorf("invalid path provided, absolute paths must be used, %s does not have prefix %s", absPath, protoPrefix)
-		}
+			if !util.PathMatchesPathElemPrefix(absPath, protoPrefix) {
+				if mustBeChildren {
+					return nil, fmt.Errorf("invalid path provided, absolute paths must be used, %s does not have prefix %s", absPath, protoPrefix)
+				}
+				continue
+			}
 
-		// make the path absolute, and a schema path.
-		pp := util.TrimGNMIPathElemPrefix(absPath, protoPrefix)
+			// make the path absolute, and a schema path.
+			pp := util.TrimGNMIPathElemPrefix(absPath, protoPrefix)
 
-		if len(pp.GetElem()) == 1 {
-			directCh[pp] = v
-		}
-		// TODO(robjs): it'd be good to have something here that tells us whether we are in
-		// a compressed schema. Potentially we should add something to the generated protobuf
-		// as a fileoption that would give us this indication.
-		if len(pp.Elem) == 2 {
-			if pp.Elem[len(pp.Elem)-2].Name == "config" || pp.Elem[len(pp.Elem)-2].Name == "state" {
+			switch directOnly {
+			case true:
+				if len(pp.GetElem()) == 1 {
+					directCh[pp] = v
+				}
+				// TODO(robjs): it'd be good to have something here that tells us whether we are in
+				// a compressed schema. Potentially we should add something to the generated protobuf
+				// as a fileoption that would give us this indication.
+				if len(pp.Elem) == 2 {
+					if pp.Elem[len(pp.Elem)-2].Name == "config" || pp.Elem[len(pp.Elem)-2].Name == "state" {
+						directCh[pp] = v
+					}
+				}
+			case false:
 				directCh[pp] = v
 			}
 		}
+		return directCh, nil
+	}
+	// It is an error if we get called with values that aren't at least indirect children at this point.
```

Comment:
What does this comment mean?

Reply:
This is just clarifying the reason for why we set the `mustBeChildren` argument to `true` here. When we are finding the children of this protobuf, we are either:

 * on the initial call -- where we expect that all paths that are supplied must be children of the root protobuf that is supplied.
 * on a recursive call to this function, at which point we have filtered the paths to find the children of the child of the root that we recursed into -- such that if we received a path that was outside of this subtree, it's an error earlier in the recursion.

I'll make the comment clearer as it doesn't quite say this :)

Reply:
Done.